### PR TITLE
Fix Bugs in Sycamore Pipeline

### PR DESCRIPTION
**Pull request by karanataryn**
Fixes bug with batched partitioning (did not wait for pdfminer before extracting tables) and allows Tables to handle empty bboxes.

Without the fix, in non-ocr mode partitioning `basic_table.pdf` will fail to include the table output in the correct format (with the bounding box as a list), and instead return it as a `BoundingBox` type. This indicates that the table extraction model never saw the input tokens to help with extraction, and hence was not able to modify it to the correct format.

### CODE REVIEW

File: lib/sycamore/sycamore/data/bbox.py
Comment (eric-anderson) [inline lib/sycamore/sycamore/data/bbox.py:164]: Suggest adding a test for this since it was a bug.

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (eric-anderson) [inline lib/sycamore/sycamore/transforms/detr_partitioner.py:564]: You can dump this tracemalloc code, it didn't turn out to be that useful.

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (eric-anderson) [inline lib/sycamore/sycamore/transforms/detr_partitioner.py:555]: also suggest adding a test to catch the pdfminer dependency for extraction (it looks like that was the problem), to make sure it doesn't regress.

Comment (eric-anderson) [reply lib/sycamore/sycamore/transforms/detr_partitioner.py:555]: Ok, I just took another look. This shouldn't be doing anything. batches is a generator, but it was consumed in the first pass.  If it's not, then we have a memory issue since you'd be keeping the entire pdf images in memory.

```
def generator(n):
    for i in range(n):
        print(f"yielding {n}")
        yield i

f = generator(10)
for i in f:
    print(f"got {i}")

print("LOOP 2")
for i in f:
    print(f"got {i}")
```

Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/detr_partitioner.py:555]: Fixed, thanks for catching this

Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/detr_partitioner.py:555]: Fixed

File: lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py
Comment (eric-anderson) [inline lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py:67]: which of these fails without the fix?  Make sure to include that in the PR description (fixing the below to whatever is the correct statement):
```
Without the fix, in non-ocr mode partitioning basic_table.pdf will fail to include the table output.
```

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (eric-anderson) [inline lib/sycamore/sycamore/transforms/detr_partitioner.py:554]: `# TODO: optimize this to make pdfminer also streamed so we can process each page in sequence without having to double-convert the document`

File: lib/sycamore/sycamore/tests/unit/transforms/compare_detr_impls.py
Comment (eric-anderson) [inline lib/sycamore/sycamore/tests/unit/transforms/compare_detr_impls.py:18]: So long as we are running both, might as well add 
`assert deep_eq(batched, sequenced)`

File: lib/sycamore/sycamore/tests/unit/data/test_bbox.py
Comment (karanataryn) [reply lib/sycamore/sycamore/tests/unit/data/test_bbox.py:11]: Good point, added the assertion

File: lib/sycamore/sycamore/tests/unit/transforms/compare_detr_impls.py
Comment (karanataryn) [reply lib/sycamore/sycamore/tests/unit/transforms/compare_detr_impls.py:18]: Done

File: lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py
Comment (karanataryn) [reply lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py:67]: Updated

File: lib/sycamore/sycamore/tests/unit/data/test_bbox.py
Comment (bsowell) [inline lib/sycamore/sycamore/tests/unit/data/test_bbox.py:39]: Do you want union here instead of union_self? I guess it doesn't matter since in each case the boxes are empty, but seems a bit weird to update bbox4 in place, and is counter to what you did with intersect below.

File: lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py
Comment (bsowell) [reply lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py:67]: Where is this comment? Did you add it to one of the tests? Hopefully I'm not looking at an old version.

File: lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py
Comment (bsowell) [inline lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py:72]: minor: Maybe this is a normal thing for mono-repos, but I wonder if it might not be safer to put these pdfs in the sycamore library test directory just to avoid surprises if someone edits crawlers. I don't feel strongly, and I'm not thrilled with making the repo even bigger, so I'll leave it to you.

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (bsowell) [inline lib/sycamore/sycamore/transforms/detr_partitioner.py:555]: Just to make sure I understand correctly, this will double-convert even in cases when you aren't using pdfminer, right? That may be fine, but we need to do some performance measurements, I think. Can we get this on one of the test instances and run the load tests + measure latency?

**Review by bsowell — APPROVED** [review]

File: lib/sycamore/sycamore/tests/unit/data/test_bbox.py
Comment (karanataryn) [reply lib/sycamore/sycamore/tests/unit/data/test_bbox.py:39]: Makes sense, have updated it

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (karanataryn) [reply lib/sycamore/sycamore/transforms/detr_partitioner.py:555]: Will work on this after merging this in

File: lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py
Comment (karanataryn) [reply lib/sycamore/sycamore/tests/unit/transforms/test_detr_partitioner.py:72]: It is cleaner, and since it allows this to be more modular, I'll copy it over to the test directory as well

**Pull request merged**
